### PR TITLE
1 feature build frontend pagination

### LINKED ISSUE #1: Feature: Build Frontend Pagination
**THIS ISSUE WAS AUTOGENERATED BY HATCHWAYS**
For this task, you will be modifying existing client code for a blogging website. Currently the website has already implemented:

- Ability to load and display blog posts. For this evaluation, the data is retrieved from a JSON file that is already included.

Please open a single pull request with the following changes:
1. Adding pagination functionality
2. Write a markdown file with answers to questions in Part 2

The expected changes are outlined in detail below.

## Part 1 - Adding pagination abilities

Currently, the UI has already exists - **please do not change the HTML structure** for this part of the task. We would like it to be implemented so that it satisfy the following functionality:

- By default, “15 per page” is selected and you will be on page 1
- When user changes “X per page” (the only options will be 15, 25, 50 and 100), it should only display at maximum that amount of blogs per page and the first page is displayed
- When user clicks on the Previous and Next arrow button, it will display it’s respective page results
- If the first and last page is the same, then previous and next arrow buttons are disabled
- If user at the beginning of the results, the previous button should be disabled
- If user at the end of the results, the next button should be disabled
- Current page always have 2 siblings (refer to screenshots below)
- Ellipses should be inserted if the first page or the last page is not a sibling of the current page
- First and last page must always be displayed
- There will always be at least 1 blog post

Please note that for this evaluation, calculations will be done on the frontend and that we are not looking to imitate an API call.
- Current page is 1, and it has 2 sibling options next to it. Ellipses before the last page.
    first
- In the middle of the pagination, with one sibling before and after. Ellipses after page 1 and before the last page.
    second
- Second last page, one sibling before and after (the last page is the same page as the next page). Ellipses after page 1.
    third
- No ellipses when there is only one page available.
    fourth

## Part 2: Written Evaluation

Please provide a markdown file with the answers to the following questions below.

The product team has decided that we want to make improvements to this application.

Questions:

1. Right now the data for the posts is coming from a json file. What changes would you have to make to the application if it came from an API? In what type of hook should you use to fetch the data and why? What other considerations would you have to make?
2. Part of this application uses the package [nanoid](https://www.npmjs.com/package/nanoid) to generate keys. What issue would this cause for generating keys in React?

### Considerations
- Please format your answers so that they are easy to digest, and **do not include any code in your pull request** related to these questions above. We will be evaluating both the quality of your answer as well as the quality of your written explanation of that answer.
- Please include a file in the root of the repo called `improvements-proposal.md` that addresses these questions.

### PR DESCRIPTION
**Pull request by justinduplain**

### CODE REVIEW

**Review by hatchy-reviewer — COMMENTED** [review]
Congrats on opening a PR! After you verify that all [GitHub actions](https://github.com/justinduplain/317773/pull/2/checks) are passing, please return to the [Hatchways dashboard](https://app.hatchways.io/app/#/dashboard/) to submit your work.

File: improvements-proposal.md [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+##### Other Considerations
+
+I would consider implementing some kind of global state management tool such as Redux. This would help to maintain state as multiple Async requests come in to play. As a bonus, these tools can also include useful features like memoization. This can make the application perform better by limiting additional identical requests to the API. Finally, it helps to seperate MVC portions of the application: the functional logic (controller) is placed into reducers and is separate from the displayed components (view).
+
+### Q2. Part of this application uses the package [nanoid](https://www.npmjs.com/package/nanoid) to generate keys. What issue would this cause for generating keys in React?
+
+#### Q2 Answer:
+
+Nanoid is not recommended for generating React keys, as each re-render generates a new nanoid, whereas React keys are supposed to stay consistent across renders. Using nanoid could cause instability in React applications. Instead, a consistent stable ID from the object should be used.
+
```
Comment (hatchy-reviewer) [inline improvements-proposal.md:35]: - You have mentioned stable ID could be used for generating keys, can you go deeper by what you mean when you say stable ID by giving some instances of it?

File: src/components/BlogList.jsx [hunk 2]
```diff
@@ -1,22 +1,47 @@
...
 import BlogPost from "./BlogPost";
 import Pagination from "./Pagination";
-import React from "react";
 import blogs from "../data/blogs.json";
 
 const PAGE_SIZES = [15, 25, 50, 100];
 
 function BlogList() {
-  const currentPaginationData = blogs.posts.slice(0, 15);
+  /* HOOKS */
```
Comment (hatchy-reviewer) [inline src/components/BlogList.jsx]: There are several unnecessary comments in the code, be careful not to [overuse comments](https://qvault.io/clean-code/code-comments/) - comments should say something that is not obvious in the code and that would provide value to future developers.

File: src/components/BlogPost.jsx [hunk 3]
```diff
@@ -11,7 +11,8 @@ function BlogPost({ author, title, excerpt }) {
           <img
             className="authorImage"
             src={`https://joeschmoe.io/api/v1/${author}`}
-            alt="Author"
+            alt={`${author} avatar`}
+            title={author}
```
Comment (hatchy-reviewer) [inline src/components/BlogPost.jsx]: These changes are not related, revert them.

File: improvements-proposal.md [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+- The function would have to account for error handling
+  - And we would possibly want to include some kind of error feedback in the UI
+
+##### Using the useEffect hook for an API Call
+
+Working with Async functions like an API call would necessitate using React's useEffect hook. The useEffect hook allows us to update the state or components once the Async function eventually returns the requested data. Additionally, useEffect can watch for future data changes and update appropriate state/props/components when that happens.
+
+##### Other Considerations
+
+I would consider implementing some kind of global state management tool such as Redux. This would help to maintain state as multiple Async requests come in to play. As a bonus, these tools can also include useful features like memoization. This can make the application perform better by limiting additional identical requests to the API. Finally, it helps to seperate MVC portions of the application: the functional logic (controller) is placed into reducers and is separate from the displayed components (view).
```
Comment (hatchy-reviewer) [inline improvements-proposal.md:22]: Great answer!
- How could you improve your current solution if the number of blog posts was really large (for example 10,000)?

File: src/components/Pagination.jsx [hunk 5]
```diff
@@ -84,7 +97,7 @@ function Pagination({
           // Do not remove the aria-label below, it is used for Hatchways automation.
           aria-label="Goto next page"
           onClick={onNext}
-          disabled={false} // change this line to disable a button.
+          disabled={currentPage === paginationRange.slice(-1)[0]} // disables next on the last page
```
Comment (hatchy-reviewer) [inline src/components/Pagination.jsx]: It would make for better memory space efficiency to avoid making unnecessary shallow copies of the "paginationRange" array. You could access the last page in the paginationRange.

File: src/hooks/usePagination.js [hunk 6]
```diff
@@ -1,14 +1,71 @@
...
-    !!!!!! ATTENTION !!!!!!
-    Please replace this comment here with a description of this hook.
-    
-  */
-  return [1, 2, 3, DOTS, 5];
+function usePagination({ currentPage, pageSize, totalCount }) {
+  //calculates the last page number
+  const lastPage = Math.ceil(totalCount / pageSize);
+  // uses a set to avoid duplicates & errors
+  const pageSet = new Set([1]);
```
Comment (hatchy-reviewer) [inline src/hooks/usePagination.js]: You could just use an array since it is very unlikely to have duplicate values. You would avoid having to convert the Set to an array there by creating more memory locations.

File: src/hooks/usePagination.js [hunk 7]
```diff
@@ -1,14 +1,71 @@
...
+ * - Adds ellipsis to gaps (ie. non-siblings from the set)
+ * - returns an array of page numbers with placed ellispsis
+ */
 
-    !!!!!! ATTENTION !!!!!!
-    Please replace this comment here with a description of this hook.
-    
-  */
-  return [1, 2, 3, DOTS, 5];
+function usePagination({ currentPage, pageSize, totalCount }) {
```
Comment (hatchy-reviewer) [inline src/hooks/usePagination.js:8]: The value returned by this hook can be memoized/cached to avoid unnecessarily running the logic every time the hook is called. The useMemo hook seems suitable.

**Review by hatchy-reviewer — COMMENTED** [review]
Hi, I'm Temesghen from Hatchways and I'll be reviewing your code today. You can re-submit your assessment by doing the following:
- respond to any comments on GitHub and address them by pushing more commits to this branch in this pull request
- when you are ready for us to look at the pull request again, go back to this page and click the "Re-submit" button for this assessment. If you don't do this step, we won't know that you want us to look at the pull request again

By re-submitting, it can drastically improve your chances to move forward to the next steps in the interview process. Here is some general feedback about the correctness of your solution:

- Your pagination has fewer page numbers when the selected page is 3.
-- Thanks Temesghen! Yes this is my fault, the example said to place ellipses between 1 and 2 in this case. But it did not make sense to me as there are not any missing pages between 1 and 2. The instructions displayed (1) ... (2) (3) (4) ... (44). In a normal situation I would verify that this is what was wanted. I'll correct this in my next submission.

File: improvements-proposal.md (hunk 1 above)
Comment (justinduplain) [reply improvements-proposal.md:35]: A 'stable ID' can be any unique and unchanging identifier. An ID for a React key should be a string that distinguishes an item from its siblings. Some examples of stable IDs are: a user ID, a database entry ID, or a post ID.

File: improvements-proposal.md (hunk 4 above)
Comment (justinduplain) [reply improvements-proposal.md:22]: > How could you improve your current solution if the number of blog posts was really large (for example 10,000)?

This would require some attention to the UI, for example: Paging through 1000+ pages of posts seems unreasonable; would we provide another way to filter or display the posts? Does ordinal pagination still make the most sense in this case, or should we consider a "load more" or "infinite scroll" style of display? Would we request the full count from the database, or only display that there are more pages than would reasonably be expected (eg. page 5 of more than 1,000 vs. page 5 of 2,127). 

Assuming that we were actually pulling these posts from a database, I would want to have access to a pagination API. This would return only a limited number of blog posts. This greatly reduces the amount of data being loaded by the client at one time and can improve performance.

File: src/components/BlogList.jsx (hunk 2 above)
Comment (justinduplain) [reply src/components/BlogList.jsx]: Thank you for the reference, I will correct this.

File: src/components/Pagination.jsx (hunk 5 above)
Comment (justinduplain) [reply src/components/Pagination.jsx]: Good point, will correct.

File: src/hooks/usePagination.js (hunk 6 above)
Comment (justinduplain) [reply src/hooks/usePagination.js]: corrected. 

> You could just use an array since it is very unlikely to have duplicate values. You would avoid having to convert the Set to an array there by creating more memory locations.

File: src/hooks/usePagination.js (hunk 7 above)
Comment (justinduplain) [reply src/hooks/usePagination.js:8]: Added memoization using useMemo hook in the pagination component.

File: src/components/BlogPost.jsx (hunk 3 above)
Comment (justinduplain) [reply src/components/BlogPost.jsx]: reverted

**Comment by justinduplain** [comment]
Thank you for the feedback, this was very helpful.

Best,
Justin Duplain
941-301-8020
LinkedIn <https://www.linkedin.com/in/justinduplain/>
Schedule A Meeting <https://calendly.com/justinduplain/30min?month=2022-06>
